File: NeetCode 150/Graphs/Walls and Gates.py
from collections import deque
from typing import List

class Solution:
    def wallsAndGates(self, rooms: List[List[int]]) -> None:
        if not rooms or not rooms[0]:
            return
        
        rows, cols = len(rooms), len(rooms[0])
        INF = 2**31 - 1
        q = deque()
        
        # 1️⃣ Push all gates into queue
        for r in range(rows):
            for c in range(cols):
                if rooms[r][c] == 0:
                    q.append((r, c))
        
        directions = [(1,0), (-1,0), (0,1), (0,-1)]
        
        # 2️⃣ BFS
        while q:
            r, c = q.popleft()
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                
                # Only update empty rooms
                if 0 <= nr < rows and 0 <= nc < cols and rooms[nr][nc] == INF:
                    rooms[nr][nc] = rooms[r][c] + 1
                    q.append((nr, nc))

File: NeedCode 150 (JAVA)/Graphs/Walls and Gates.java
import java.util.*;

class Solution {

    public void wallsAndGates(int[][] rooms) {

        if (rooms == null || rooms.length == 0)
            return;

        int m = rooms.length;
        int n = rooms[0].length;

        Queue<int[]> queue = new LinkedList<>();

        // Add all gates to the queue
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (rooms[i][j] == 0) {
                    queue.offer(new int[]{i, j});
                }
            }
        }

        int[][] dirs = {
            {1, 0},
            {-1, 0},
            {0, 1},
            {0, -1}
        };

        while (!queue.isEmpty()) {

            int[] curr = queue.poll();
            int r = curr[0];
            int c = curr[1];

            for (int[] d : dirs) {

                int nr = r + d[0];
                int nc = c + d[1];

                // Check boundaries
                if (nr < 0 || nr >= m || nc < 0 || nc >= n)
                    continue;

                // Visit only empty rooms
                if (rooms[nr][nc] != Integer.MAX_VALUE)
                    continue;

                rooms[nr][nc] = rooms[r][c] + 1;
                queue.offer(new int[]{nr, nc});
            }
        }
    }
}
